core/src/verify.rs

```diff
@@ -199,52 +213,36 @@ impl CommitSequenceVerifier {
         if rs.genesis_info != self.reserved_state.genesis_info {
             return Err(Error::InvalidArgument("genesis_info changes".to_string()));
         }
-        // Check that the newly added (if exists) `Member::name` is unique.
-        let existing_names = self
-            .reserved_state
-            .members
-            .iter()
-            .map(|m| &m.name)
-            .collect::<HashSet<_>>();
+        // Check that `Member::name` and `Member::public_key` are unique.
         for member in &rs.members {
-            if !existing_names.contains(&member.name)
-                && rs.members.iter().filter(|m| m.name == member.name).count() > 1
-            {
+            if rs.members.iter().filter(|m| m.name == member.name).count() > 1 {
                 return Err(Error::InvalidArgument(format!(
                     "member name '{}' already exists",
                     member.name
                 )));
             }
-        }
-        // Check that `member` monotonically increases (refer to `Member::expelled`).
-        let existing_members = self
-            .reserved_state
-            .members
-            .iter()
-            .map(|m| m.public_key.clone())
-            .collect::<HashSet<_>>();
-        for member in &rs.members {
-            if existing_members.contains(&member.public_key) && member.expelled {
+            if rs
+                .members
+                .iter()
+                .filter(|m| m.public_key == member.public_key)
+                .count()
+                > 1
+            {
                 return Err(Error::InvalidArgument(format!(
-                    "member '{}' cannot be expelled",
+                    "the public key of '{}' already exists",
                     member.name
                 )));
             }
         }
-        // Check that the delegation state doesn't change.
-        if rs.members.iter().any(|m| m.governance_delegatee.is_some())
-            && rs.members != self.reserved_state.members
-        {
-            return Err(Error::InvalidArgument(
-                "governance_delegatee cannot be changed".to_string(),
-            ));
-        }
-        if rs.members.iter().any(|m| m.consensus_delegatee.is_some())
-            && rs.members != self.reserved_state.members
-        {
-            return Err(Error::InvalidArgument(
-                "consensus_delegatee cannot be changed".to_string(),
-            ));
+        // Check that `member` monotonically increases (refer to `Member::expelled`).
```

Comment:
This is just `nit` personal opinion, but it would be nice to add a comment to describe the meaning of `monotonically increase`, like for example but not limited to: `// Once a member is added, it cannot be removed, even if it is expelled.`